### DISCUSSION TITLE
Should inline options on a `CodeBlock` cause it to render as a `Para`?

### DISCUSSION THREAD

**Discussion by coatless** [quarto-dev/quarto-cli#7631] [Q&A] [answered]
### Description

I'm working on exploring a bit more as it relates to custom cell engines, e.g. `custom-language`. Part of exploration is regarding capturing cell options since they are not stored in the [`CodeBlock`](https://pandoc.org/lua-filters.html#type-codeblock) (via `.attributes`/`.classes`/ ...).

I am perplexed by "inline options" or options declared immediately after the engine. For example, we have:

````md
---
title: "Cell Options"
format: native
---

### Inline

```{custom-r, option1=value1, option2=value2}
# code here

1 + 1
```

### Hashpipe

```{custom-r}
#| option1: value1
#| option2: value2

# code here
1 + 1
```
````

The rendered output in HTML would be: 

<img width="320" alt="Rendered comparison of inline vs. hashpipe" src="https://github.com/quarto-dev/quarto-cli/assets/833642/d78e64ab-9ad4-49a0-ae8c-69dc106ac601">

Switching over to the native structure, the inline code cell options seem to be converting the `CodeBlock` element  into a `Para` structure followed by inline text:

```
Pandoc
  Meta
    { unMeta =
        fromList
          [ ( "title"
            , MetaInlines [ Str "Cell" , Space , Str "Options" ]
            )
          ]
    }
  [ Header 3 ( "inline" , [] , [] ) [ Str "Inline" ]
  , Para
      [ Str "```{custom-r,"
      , Space
      , Str "option1=value1,"
      , Space
      , Str "option2=value2}"
      , SoftBreak
      , Str "#"
      , Space
      , Str "code"
      , Space
      , Str "here"
      ]
  , Para [ Str "1" , Space , Str "+" , Space , Str "1" ]
  , CodeBlock
      ( "" , [] , [] )
      "\n### Hashpipe\n\n```{custom-r}\n#| option1: value1\n#| option2: value2\n\n# code here\n1 + 1"
  ]
```

Is this output correct? Shouldn't it be passed down as a `CodeBlock` instead of a `Para`?

**Comment by mcanouil** [comment]
What if a you use YAML style which is the style used and recommended in Quarto?

There's might be some knitr specific behaviour here (either directly in knitr or the in R codebase in Quarto)

**Reply by coatless** [reply]
So, the "hashpipe" style or YAML style shown after "inline" works great. I've already implemented a custom per-code cell parser.

The issue above crept up as I had someone ask in an e-mail about why the "traditional rmarkdown" syntax wasn't working. 

I'm not sold that it's a `knitr` issue since if I specify `engine: jupyter`, I'll end up with a similar `native` & `html` format outputs. 

## `engine: knitr`

<details>
<summary>Run with `engine: knitr` to native format</summary>

```sh
quarto preview /workspaces/codecelloptions/example.qmd --no-browser --no-watch-inputs
```

```sh
processing file: example.qmd
1/2                  
2/2 [unnamed-chunk-1]
output file: example.knit.md

Warning message:
In grSoftVersion() :
  unable to load shared object '/usr/local/lib/R/modules//R_X11.so':
  libXt.so.6: cannot open shared object file: No such file or directory
pandoc 
  to: native
  output-file: example.native
  standalone: true
  default-image-extension: png
  
metadata
  title: Cell Options
  
Output created: example.native

Watching files for changes
Browse at http://localhost:4230/
```

````md
Pandoc
  Meta
    { unMeta =
        fromList
          [ ( "title"
            , MetaInlines [ Str "Cell" , Space , Str "Options" ]
            )
          ]
    }
  [ Header 3 ( "inline" , [] , [] ) [ Str "Inline" ]
  , Para
      [ Str "```{custom-r,"
      , Space
      , Str "option1=value1,"
      , Space
      , Str "option2=value2}"
      , SoftBreak
      , Str "#"
      , Space
      , Str "code"
      , Space
      , Str "here"
      ]
  , Para [ Str "1" , Space , Str "+" , Space , Str "1" ]
  , CodeBlock
      ( "" , [] , [] )
      "\n### Hashpipe\n\n```{custom-r}\n#| option1: value1\n#| option2: value2\n\n# code here\n1 + 1"
  ]
````
</details>

## `engine: jupyter`

<details>
<summary>Run with `engine: jupyter` to native format</summary>

```sh
quarto preview /workspaces/codecelloptions/example.qmd --no-browser --no-watch-inputs
```

```sh
Starting python3 kernel...Done

Executing 'example.ipynb'

pandoc 
  to: native
  output-file: example.native
  standalone: true
  default-image-extension: png
  
metadata
  title: Cell Options
  
Output created: example.native

Watching files for changes
Browse at http://localhost:7938/
```

**Output**

````md
Pandoc
  Meta
    { unMeta =
        fromList
          [ ( "title"
            , MetaInlines [ Str "Cell" , Space , Str "Options" ]
            )
          ]
    }
  [ Header 3 ( "inline" , [] , [] ) [ Str "Inline" ]
  , Para
      [ Str "```{custom-r,"
      , Space
      , Str "option1=value1,"
      , Space
      , Str "option2=value2}"
      , SoftBreak
      , Str "#"
      , Space
      , Str "code"
      , Space
      , Str "here"
      ]
  , Para [ Str "1" , Space , Str "+" , Space , Str "1" ]
  , CodeBlock
      ( "" , [] , [] )
      "\n### Hashpipe\n\n```{custom-r}\n#| option1: value1\n#| option2: value2\n\n# code here\n1 + 1"
  ]
````

</details>

**Reply by mcanouil** [reply]
I would really recommend to tell Quarto users to use YAML style when using Quarto.

This being said, did you try in regular rmarkdown/knitr documents?

~~Also the inline syntax is not something that exists in Jupyter, so unless there is some code in Quarto to parse Jupyter cell as knitr (which would really be a big surprise), I would not expect the inline R/knitr syntax to work outside of R/knitr.~~

Edit: I noticed afterwards you implied that YAML did not work for both engines.
I recall a discussion about passing custom cell options but can't find it again for now.

**Comment by cderv** [comment, marked as answer]
> I am perplexed by "inline options" or options declared immediately after the engine. For example, we have:

This  is **knitr** only - It has nothing to do with Quarto. Quarto has no knowledge of inline options. 

This syntax with options after engine, that was used in R Markdown ecosystem, is the historical **knitr** way of passing options. 

When the engine is known by **knitr**, then what is after the engine is parsed as options provided with R syntax.  

Quarto does know about YAML options syntax for validation and auto completion, but regarding cell evaluation, it will defer to **knitr** for options parsing (the YAML parsing code for those options are in **knitr**). For Jupyter, it will handle it though. 

> I'm not sold that it's a knitr issue since if I specify engine: jupyter, I'll end up with a similar native & html format outputs.

There is no really issue here. You get the same results with knitr or jupyter because `custom-r` is not a known engine, neither by **knitr** or Jupyter. So the codeblock is unprocessed and pass as-is for Quarto to convert it by Pandoc. YAML or not YAML syntax. 

And here, your observation is mixed up because you are putting the two cells in one document. The cell with inline option causing the other one to have parsing problem. 

With Pandoc reader when encountering the cell with inline options, the native representation is what you see (Para), because this is not valid CodeBlock syntax in Pandoc's Markdown. So pandoc parser does not see it as a CodeBlock. 

Let's me try to be clearer with commenting your code 

````markdown
### Inline

```{custom-r, option1=value1, option2=value2} <-- `custom-r` is unknown, so passed as is by engine to quarto and then to pandoc and this is not valid Pandoc's markdown so Para
# code here

1 + 1
``` <-- This is seen by Pandoc as start of a code block

### Hashpipe

```{custom-r}
#| option1: value1
#| option2: value2

# code here
1 + 1
``` <-- this is matched as the end of the code block. Anything in between being part of it. 
```` 

If you do try to convert only with HAspipe
````markdown
---
title: "Cell Options"
format: native
---

### Hashpipe

```{custom-r}
#| option1: value1
#| option2: value2

# code here
1 + 1
```

````

Then you'll get something more clear
````
Pandoc
  Meta
    { unMeta =
        fromList
          [ ( "title"
            , MetaInlines [ Str "Cell" , Space , Str "Options" ]
            )
          ]
    }
  [ Header 3 ( "hashpipe" , [] , [] ) [ Str "Hashpipe" ]
  , CodeBlock
      ( "" , [ "{custom-r}" ] , [] )
      "#| option1: value1\n#| option2: value2\n\n# code here\n1 + 1"
  ]
````

This representation is I believe what shinylive relies one. 

I hope this helps understand. 

Though, let me say you are playing with edge case and internals. Quarto does not have yet an extension mechanism for Custom engines. So anything you can do with it in Lua could be broken when we do decide to add some pre engine processing, or a mechanism for custom engine. 

Though webR extension, and shinylive are using part of it, so obviously we'll make sure they don't break. 

Anyhow, I hope my explanation clarifies. Feel free to ask more if still some questions.

**Reply by coatless** [reply]
@cderv thank you for the detailed explanation!

The explanation was a great help in understanding what was happening and why I was seeing oddness in the `native` format. Moreover, I wasn't aware the inline option was specific to `knitr`.

One quick follow-up, I suppose this means that registered cell engines with `knitr` are going from: 

````md
```{custom, option1=value1, option2=value2}
```
````
To: 
````md
```{.custom .value1}
# Evaluated code
```
**Custom Option 2** 
````

Before being passed into Pandoc? This would (roughly) match with the description of [8.5.2.3 Extension: fenced_code_attributes](https://pandoc.org/chunkedhtml-demo/8.5-verbatim-code-blocks.html)

---

On the note of edge cases and internals, I'm aware that "here be dragons" 🐉 .  Part of the reason for the question was the outputs just didn't make sense. Plus, for some reason, I thought inline support was ubiquitous and wanted to make sure the new cell options system I'm working on supported it. Now I know...

**Reply by cderv** [reply]
> One quick follow-up, I suppose this means that registered cell engines with knitr are going from:

Not exactly. 

First inline options or in-chunk YAML options are the same for **knitr** - they both will be parsed as options for the engine. Using inline chunk options syntax has no relation to Pandoc's syntax. 

If you were to register `custom` engine, then **knitr** would know about it and process the options as part of that engine. No matter if provided as inline or in-chunk yaml. 

Also, **knitr** won't specifically set some classes or attributes on the output code block. `attr.*` or `class.*` knitr options are usually for that purpose. But you can't do direct relationship on **knitr** syntax and its processing and Pandoc's syntax. 

Though in Quarto context, unknown options from **knitr** and Quarto (so could apply to a new custom engine), engine options will be set as code attributes - not classes. I would not rely on that as this is not a feature per se but an internal behavior useful in some case. Just wanted to be precised in case you happen to do some testing and notice it.

<details>
<summary>Just showing those examples to help understand</summary>

````markdown
---
title: "Untitled"
format: native
keep-md: true
---


```{r, custom="nothing"}
# empty 
```

```{r}
#| custom: "nothing"
# empty
```

```{r}
#| include: false
knitr::knit_engines$set(custom = function(options) {
  options$lang <- "default"
  options$code <- "From custom engine"
  knitr::knit_engines$get("cat")(options) 
  })
```

```{custom}
#| custom: nothing

```

````

````
Pandoc
  Meta
    { unMeta =
        fromList [ ( "title" , MetaInlines [ Str "Untitled" ] ) ]
    }
  [ Div
      ( "" , [ "cell" ] , [ ( "custom" , "nothing" ) ] )
      [ CodeBlock ( "" , [ "r" , "cell-code" ] , [] ) "# empty " ]
  , Div
      ( "" , [ "cell" ] , [ ( "custom" , "nothing" ) ] )
      [ CodeBlock ( "" , [ "r" , "cell-code" ] , [] ) "# empty" ]
  , Div
      ( "" , [ "cell" ] , [ ( "custom" , "nothing" ) ] )
      [ CodeBlock
          ( "" , [ "default" , "cell-code" ] , [] )
          "From custom engine"
      ]
  ]
````

See that 
* engine worked and inserted content
* `custom` options was put on cell because unknown option
* `custom` engine was processed 

</details>

Hope it helps

**Reply by cscheid** [reply]
To add to the complication of the scenario: Pandoc 3 does _not_ natively support the following syntax

````
```{python}
print("This is Python code")
```
````

See:

````
$ pandoc3 -f markdown -t
native
```{python}
print("This is Python code")
```
^D
[ Para
    [ Code
        ( "" , [] , [] ) "{python} print(\"This is Python code\")"
    ]
]
````

In pandoc 2, you get:

````
$ pandoc2 -f markdown -t native
```{python}
print("This is Python code")
```
[ CodeBlock
    ( "" , [ "{python}" ] , [] )
    "print(\"This is Python code\")"
]
````

In order to parse this correctly, quarto needs to use a _custom reader_ (see `qmd-reader.lua`, and welcome to the wonderful world of Lua's PEG parsers!). I wrote that, and we cannot support syntax any more complicated that what we have: the code is already very subtle and brittle. Any deviation from Pandoc 3's parsing needs a lot of justification before we can take in. The `{lang}` syntax is of course central to the rmarkdown/quarto ecosystem and so our hand was forced. But we're not going to extend this any further, unless something truly unexpected comes up.
